c_src/notif_set_tlh_htable.c

```diff
@@ -0,0 +1,245 @@
+        new_key_int_added = true; //necessary for propser cleanup on error
+    }
+    else {
+        htable_key_free(key_int); //because did not put it in the htable
+    }
+
+    // add the filename to the level2 hash table of this wd
+    key_str = htable_key_str(msg->name);
+    CHECK_SUCCESS(key_str);
+    struct htable_ret put_ret = htable_put(lev2_tab, key_str, 0);
+    CHECK_SUCCESS(!put_ret.err_flag);
+    if(!put_ret.ok) { //filename is already in the set
+        htable_key_free(key_str); //because did not put it in the htable
+        return false;
+    }
+
+    tab->len++;
+    return true;
+
+fn_fail:
+    if(new_key_int_added) {
+        htable_del(&tab->lev1, key_int);
+    }
+    else {
+        htable_key_free(key_int);
+    }
+    htable_key_free(key_str);
+    *err_flag = 1;
+    return false;
+}
+
+bool notif_set_find(const struct Notif_set *tab, const struct Message *msg, int *err_flag)
+{
+    *err_flag = 0;
+    struct Key *key = htable_key_int(msg->wd);
+    CHECK_SUCCESS(key);
+    struct htable_ret get_ret = htable_get(&tab->lev1, key);
+    htable_key_free(key);
+    if(!get_ret.ok) {
+        return false;
+    }
+    key = htable_key_str(msg->name);
+    CHECK_SUCCESS(key);
+    get_ret = htable_get((struct htable*) get_ret.value, key);
+    htable_key_free(key);
+    if(!get_ret.ok) {
+        return false;
+    }
+
+    return true;
+
+fn_fail:
+    *err_flag = 1;
+    return false;
+}
+
+struct Notif_set_itr *notif_set_itr_init(struct Notif_set *tab)
+{
+    struct Notif_set_itr *itr = calloc(1, sizeof(*itr));
+    CHECK_SUCCESS(itr);
+    if(tab != NULL) {
+        itr->tab = tab;
+        itr->itr1 = htable_next(&tab->lev1, NULL);
+    }
+
+    return itr;
+
+fn_fail:
+    return NULL;
+}
+
+void notif_set_itr_destroy(struct Notif_set_itr *itr)
+{
+    free(itr);
+}
+
+bool notif_set_itr_next(struct Notif_set_itr *itr, struct Message *msg)
+{
+    if(itr->itr1 == NULL) {
+        return false;
+    }
+    // Advance the second-level iterator first
+    if((itr->itr2 = htable_next((struct htable *) itr->itr1->value, itr->itr2))) {
+        assert(itr->itr1->key->n == sizeof(uint64_t));
+        msg->wd = *((uint64_t*) itr->itr1->key->bytes); //Really don't like this
+        memcpy(msg->name, itr->itr2->key->bytes, itr->itr2->key->n);
+        return true;
+    }
+    // If second level is exhausted, advance the iterator in the first level
+    if((itr->itr1 = htable_next(&itr->tab->lev1, itr->itr1))) {
+        // Call recursively to iterate in the second level
+        return notif_set_itr_next(itr, msg);
+    }
+
+    return false;
+}
+
+bool notif_set_itr_remove(const struct Notif_set_itr *itr)
+{
+    // First, attempt to remove from the second-level table
+    if(itr->itr2) {
+        if(htable_key_free((struct Key*) itr->itr2->key)) {
+            itr->itr2->key = NULL;
+            ((struct htable*)(itr->itr1->value))->len--;
+            // TODO: We could destroy the level2 htable and remove
```

Comment:
This is a good justification for not destroying the level2 table here. I would leave the comment, but remove the TODO.